shell/platform/darwin/ios/framework/Source/profiling_metrics.h

```diff
@@ -0,0 +1,62 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_SHELL_PLATFORM_DARWIN_IOS_FRAMEWORK_SOURCE_PROFILING_METRICS_H_
+#define FLUTTER_SHELL_PLATFORM_DARWIN_IOS_FRAMEWORK_SOURCE_PROFILING_METRICS_H_
+
+#include <mach/mach.h>
+
+#include <cassert>
+#include <optional>
+
+#include "flutter/fml/logging.h"
+#include "flutter/shell/profiling/profiler.h"
+
+namespace flutter {
+
+std::optional<flutter::CpuUsageInfo> CpuUsageIOS() {
+  kern_return_t kernel_return_code;
+  thread_array_t threads;
+  mach_msg_type_number_t thread_count;
+
+  // Get threads in the task
+  kernel_return_code = task_threads(mach_task_self(), &threads, &thread_count);
+  if (kernel_return_code != KERN_SUCCESS) {
+    FML_LOG(ERROR) << "Error retrieving task information: "
+                   << mach_error_string(kernel_return_code);
+    return std::nullopt;
+  }
+
+  double total_cpu_usage = 0.0;
+
+  // Add CPU usage for each thread.
+  for (mach_msg_type_number_t i = 0; i < thread_count; i++) {
+    thread_basic_info_data_t basic_thread_info;
+    mach_msg_type_number_t thread_info_count = THREAD_BASIC_INFO_COUNT;
+    kernel_return_code =
+        thread_info(threads[i], THREAD_BASIC_INFO,
+                    (thread_info_t)&basic_thread_info, &thread_info_count);
+    if (kernel_return_code != KERN_SUCCESS) {
+      FML_LOG(ERROR) << "Error retrieving thread information: "
+                     << mach_error_string(kernel_return_code);
+      return std::nullopt;
+    }
+    const double current_thread_cpu_usage =
+        basic_thread_info.cpu_usage / static_cast<float>(TH_USAGE_SCALE);
+    total_cpu_usage += current_thread_cpu_usage;
```

Comment:
I wonder if we also have to consider the difference between big (high performance) cores and small (high efficiency) cores? A 100% big core thread probably consumes more energy than a 100% small core thread (https://en.wikipedia.org/wiki/Apple_A12) so we might have to normalize accordingly. So far I couldn't Google any code that treats different cores separately so I'm Ok with this approach.

Reply:
very interesting point. I haven't considered this.